src/frontend/deployment_details.go

```diff
@@ -0,0 +1,67 @@
+package main
+
+import (
+	"net/http"
+	"os"
+	"sync"
+
+	"cloud.google.com/go/compute/metadata"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	DetailsAreNotLoaded = 0
+	DetailsAreLoading   = 1
+	DetailsAreLoaded    = 2
+)
+
+var deploymentDetailsMap = make(map[string]string)
+var detailsLoadingStatus = DetailsAreNotLoaded
+var detailsLoadingStatusMutex sync.Mutex // Ensures only 1 HTTP request loads deployment details.
+
+func loadDeploymentDetails(httpRequest *http.Request) map[string]string {
+	var metaServerClient = metadata.NewClient(&http.Client{})
+
+	// The use of httpRequest here lets us to see HTTP request info in the logs.
+	var log = httpRequest.Context().Value(ctxKeyLog{}).(logrus.FieldLogger)
+
+	podHostname, err := os.Hostname()
+	if err != nil {
+		log.Error("Failed to fetch the hostname for the Pod", err)
+	}
+
+	podCluster, err := metaServerClient.InstanceAttributeValue("cluster-name")
+	if err != nil {
+		log.Error("Failed to fetch the name of the cluster in which the pod is running", err)
+	}
+
+	podZone, err := metaServerClient.Zone()
+	if err != nil {
+		log.Error("Failed to fetch the Zone of the node where the pod is scheduled", err)
+	}
+
+	deploymentDetailsMap["HOSTNAME"] = podHostname
+	deploymentDetailsMap["CLUSTERNAME"] = podCluster
+	deploymentDetailsMap["ZONE"] = podZone
```

Comment:
Why don't we just get this information once, and only once ever? We can do this from an `init()` function, with a package variable that we can query as many times as we want.

```
// deployment_details.go
package main

import (...)

var deploymentDetails map[string]string

func init() {
    // get the deployment details from the meta server

    deploymentDetails = map[string]string {
        "HOSTNAME": podHostname,
        "CLUSTERNAME": podCluster,
        "ZONE": podZone,
    }
}
```

Then from the handlers:
```
if err := templates.ExecuteTemplate(w, "home", map[string]interface{}{
    . . .
    . . .
    "deploymentDetails": deploymentDetails,
}
```

That would mean no mutex, no enums, no race conditions, no "loading / not loading", and reduce complexity.

Reply:
If the meta server is not available or reachable (e.g. if you're deploying locally), then you would leave the map blank, and that should be taken care of seamlessly by the HTML template with the "if deploymentDetails" bit.

Reply:
Great suggestion! I'm glad you took a look at my changes.
I'm now loading the deployment details from `func init()`.
I was able to get rid of the following complexities: 
* mutex
* race conditions
* enums

However, I wasn't able to get rid of the "no 'loading / not loading'" stuff. This is because we don't want the loading of details to block the main thread in non-GCP deployments (as described in #685).

Reply:
I wonder, if we should still account for temporary unreachability of the `metadata server`. I thought the previous mutex approach took care of all possibilities. The current change would render the details never loaded if the first attempt on init fails. 

Just a thought. We could still see how this works and add it as an improvement if need be.

Reply:
That's a fair point, though, in what scenario would the meta server to be unavailable? I thought maybe if the GCE API (https://status.cloud.google.com/) is down completely, but the meta server is internal, no? The SLA for any arbitrary single-zone cluster is `>= 99.5%`, so I'm imagining the meta server to be much higher, though that is solely intuition.

I think that's a very low probability, but even so mutex seems overkill. We could do periodic retries in the go routine until we get a result, for example (though I think that's still overkill).

I think it's safe to leave it as-is, but I'm also opened to adding retries (taking into account that it may retry forever for non-GCP instances).